### PR TITLE
[ISSUE #1751]Fix bug in MessageClientIDSetter

### PR DESCRIPTION
**Pull request by rushsky518**
pid in msgId is erased

### CODE REVIEW

File: common/src/main/java/org/apache/rocketmq/common/message/MessageClientIDSetter.java [hunk 1]
```diff
@@ -37,13 +37,10 @@
...
         }
-        LEN = ip.length + 2 + 4 + 4 + 2;
-        ByteBuffer tempBuffer = ByteBuffer.allocate(ip.length + 2 + 4);
-        tempBuffer.position(0);
+        LEN = ip.length + 4 + 4 + 4 + 2;
+        ByteBuffer tempBuffer = ByteBuffer.allocate(ip.length + 4 + 4);
         tempBuffer.put(ip);
-        tempBuffer.position(ip.length);
         tempBuffer.putInt(UtilAll.getPid());
-        tempBuffer.position(ip.length + 2);
```
Comment (duhenglucky) [inline common/src/main/java/org/apache/rocketmq/common/message/MessageClientIDSetter.java]: pid is short，so maybe putShort can resolve this issue，and this indeed a bug, ip.length + 2 will keep the lower 2 bytes.

**Comment by duhenglucky** [comment]
@rushsky518 thanks for your contribution, this bug will cause MsgId repeat, and could you pull an issue first? we will merge this PR ASAP.

File: common/src/main/java/org/apache/rocketmq/common/message/MessageClientIDSetter.java (hunk 1 above)
Comment (rushsky518) [reply common/src/main/java/org/apache/rocketmq/common/message/MessageClientIDSetter.java]: "The pid_t data type is a signed integer"
https://www.gnu.org/software/libc/manual/html_node/Process-Identification.html

Comment (rushsky518) [reply common/src/main/java/org/apache/rocketmq/common/message/MessageClientIDSetter.java]: https://unix.stackexchange.com/questions/16883/what-is-the-maximum-value-of-the-process-id

**Comment by rushsky518** [comment]
> @rushsky518 thanks for your contribution, this bug will cause MsgId repeat, and could you pull an issue first? we will merge this PR ASAP.

https://github.com/apache/rocketmq/issues/1751

File: common/src/main/java/org/apache/rocketmq/common/message/MessageClientIDSetter.java (hunk 1 above)
Comment (rushsky518) [reply common/src/main/java/org/apache/rocketmq/common/message/MessageClientIDSetter.java]: issue is here：https://github.com/apache/rocketmq/issues/1751

Comment (rushsky518) [reply common/src/main/java/org/apache/rocketmq/common/message/MessageClientIDSetter.java]: and there is a redundant "org.apache.rocketmq.common.MixAll#getPID"

**Comment by 2259289435** [comment]
pid是short值，原先的写法putInt在后退两2，抹掉了低位，导致单机多jvm时pid的隔离效果消失，msgid会产生重复的bug

所以直接putshort（（short）pid）就可以修复这个问题了，后续都不用动，也符合原先pid 占用 2 的本意。
不需要修改成占用 4



2259289435@qq.com
 
From: rushsky518
Date: 2020-02-12 13:38
To: apache/rocketmq
CC: Subscribed
Subject: Re: [apache/rocketmq] Fix bug in MessageClientIDSetter (#1758)
@rushsky518 commented on this pull request.


In common/src/main/java/org/apache/rocketmq/common/message/MessageClientIDSetter.java:
>          tempBuffer.putInt(UtilAll.getPid());
-        tempBuffer.position(ip.length + 2);

issue is here：#1751

File: common/src/main/java/org/apache/rocketmq/common/message/MessageClientIDSetter.java [hunk 2]
```diff
@@ -107,6 +104,15 @@ public static String getIPStrFromID(String msgID) {
         return result;
     }
 
+    public static short getPidFromID(String msgID) {
```
Comment (duhenglucky) [inline common/src/main/java/org/apache/rocketmq/common/message/MessageClientIDSetter.java]: it seems this method repeat with org.apache.rocketmq.common.message.MessageClientIDSetter#getIPFromID

**Pull request merged**